xpti/include/xpti_data_types.h

```diff
@@ -0,0 +1,484 @@
+                                                    
+typedef std::atomic<bool>                            safe_flag_t;
+typedef std::atomic<uint64_t>                        safe_uint64_t;
+typedef std::atomic<uint32_t>                        safe_uint32_t;
+typedef std::atomic<uint16_t>                        safe_uint16_t;
+typedef std::atomic<int64_t>                         safe_int64_t;
+typedef std::atomic<int32_t>                         safe_int32_t;
+typedef std::atomic<int16_t>                         safe_int16_t;
+typedef std::unordered_map<string_id_t, string_id_t> metadata_t;
+
+#define XPTI_EVENT(val)             xpti::event_type_t(val)
+#define XPTI_TRACE_POINT_BEGIN(val) xpti::trace_point_t(val<<1 | 0)
+#define XPTI_TRACE_POINT_END(val)   xpti::trace_point_t(val<<1 | 1)
+
+#define XPTI_PACK08_RET16(value1, value2) ((value1 <<  8) | value2)
+#define XPTI_PACK16_RET32(value1, value2) ((value1 << 16) | value2)
+#define XPTI_PACK32_RET64(value1, value2) (((uint64_t)value1 << 32) | value2)
+
+/// @brief Payload data structure that is optional for trace point callback API
+/// @details The payload structure, if determined at compile time, can deliver
+/// the source association of various parallel constructs defined by the language.
+/// In the case it is defined, a lookup table will provide the association from 
+/// a kernel/lambda (address) to a payload and the same address to a unique ID 
+/// created at runtime.
+///
+/// All instances of a kernel will be associated with the same unique ID through
+/// the lifetime of an object. The hash maps that will be maintained would be:
+/// # [unique_id]->[payload]
+/// # [kernel address]->[unique_id]
+/// 
+/// Unique_id MUST be propagated downstream to the OpenCL runtime to ensure the 
+/// associations back to the sources. This requires elp from the compiler front-end.
+///
+typedef struct payload_t
+{
+    string_id_t name_sid;       ///< String ID of the function/kernel name; only valid if 'name' is valid
+    string_id_t source_file_sid;///< String ID of the source file name
+    const char *name;           ///< Name of the trace point; graph, algorithm, lock names, for example. 
+    const char *source_file;    ///< Absolute path of the source file; may have to to be unicode string
+    uint32_t    line_no,        ///< Line number information to correlate the trace point
+                column_no;      ///< For a complex statement, column number may be needed to resolve the trace point
+    const void *code_ptr_va;    ///< Kernel/lamba/function address
+    uint64_t    internal;       ///< Internal bookkeeping slot - do not change. 
+    uint64_t    flags;       ///< Flags indicating whether string name, codepointer, source file and hash values are available
+
+    //  Initialize the structure to defaults
+    //  that mark it as an invalid payload
+    //
+    payload_t() {
+        flags           = 0;          ///< Mark the payload structure as invalid
```

Comment:
Can you remove the default constructor and move directly all the initializations to the member definitions?

Reply:
Will group this with the API changes planned.